### PR TITLE
mertrics: report region flow

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
@siddontang @BusyJay PTAL

### CODE REVIEW

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -1372,6 +1375,7 @@ impl Peer {
         }
 
         metrics.keys_written += ctx.wb.count() as u64;
+        self.bytes_written += ctx.wb.data_size() as u64;
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Why not put it into RaftReadyMetrics？And we may need to count the size of entries written into rocksdb too.

Comment (zhangjinpeng1987) [inline src/raftstore/store/peer.rs]: If we put it into RaftReadyMetrics , each region need a variable.

Comment (BusyJay) [inline src/raftstore/store/peer.rs]: I don't think so. RaftReadyMetrics is stored in Store.

Comment (siddontang) [inline src/raftstore/store/peer.rs]: I think it is ok here, we may report this to PD later.

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -1146,6 +1147,28 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+    }
+
+    fn on_report_write_bytes(&mut self, event_loop: &mut EventLoop<Self>) {
+        for (_, peer) in &mut self.region_peers {
+            if !peer.is_leader() {
+                peer.bytes_written = 0;
+                continue;
+            }
+
+            REGION_BYTES_WRITTEN_HISTOGRAM.observe(peer.bytes_written as f64);
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: LocalHistogram not export yet.

Comment (siddontang) [inline src/raftstore/store/store.rs]: merged, you can use it now.

File: src/raftstore/store/msg.rs [hunk 3]
```diff
@@ -34,6 +34,7 @@ pub enum Tick {
     SnapGc,
     CompactLockCf,
     ConsistencyCheck,
+    ReportWriteBytes,
```
Comment (siddontang) [inline src/raftstore/store/msg.rs]: it is not only for Bytes now.

File: src/raftstore/store/config.rs [hunk 4]
```diff
@@ -113,6 +115,8 @@ pub struct Config {
 
     // Interval (ms) to check region whether the data is consistent.
     pub consistency_check_tick_interval: u64,
+
+    pub report_write_bytes_interval: u64,
```
Comment (siddontang) [inline src/raftstore/store/config.rs]: use a short value in test so we can cover this feature.

File: src/raftstore/store/store.rs [hunk 5]
```diff
@@ -1146,6 +1153,33 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
 
+    fn register_report_region_flow_tick(&self, event_loop: &mut EventLoop<Self>) {
+        if let Err(e) = register_timer(event_loop,
+                                       Tick::ReportRegionFlow,
+                                       self.cfg.report_region_flow_interval) {
+            error!("{} register raft gc log tick err: {:?}", self.tag, e);
+        };
+    }
+
+    fn on_report_region_flow(&mut self, event_loop: &mut EventLoop<Self>) {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: please cover it in test

**Comment by siddontang** [comment]
PTAL @BusyJay @hhkbp2

**Comment by hhkbp2** [comment]
LGTM

File: tests/storage/test_raft_storage.rs [hunk 6]
```diff
@@ -28,6 +28,7 @@ use super::sync_storage::SyncStorage;
 
 fn new_raft_storage() -> (Cluster<ServerCluster>, SyncStorage, Context) {
     let mut cluster = new_server_cluster_with_cfs(0, 1, ALL_CFS);
+    cluster.cfg.raft_store.report_region_flow_interval = 100; // 100ms
```
Comment (siddontang) [inline tests/storage/test_raft_storage.rs]: initialize this in new_store_cfg directly.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/raftstore/store/local_metrics.rs [hunk 7]
```diff
@@ -23,7 +23,7 @@ pub struct RaftReadyMetrics {
     pub commit: u64,
     pub append: u64,
     pub snapshot: u64,
-    pub keys_written: u64,
+    pub store_keys_written: u64,
```
Comment (siddontang) [inline src/raftstore/store/local_metrics.rs]: written_keys,written_bytes may be better? 

/cc @BusyJay @hhkbp2

Comment (hhkbp2) [inline src/raftstore/store/local_metrics.rs]:
```
peer.bytes_written peer.keys_written
store.region_bytes_written store. region_keys_written
metrics.store_keys_written
```
The prefix levels look fine to me.

**Comment by zhangjinpeng1987** [comment]
PING

File: src/raftstore/store/store.rs [hunk 8]
```diff
@@ -1144,6 +1151,33 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+        for (_, peer) in &mut self.region_peers {
+            if !peer.is_leader() {
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: If peer steps down, all records will be lost.

File: src/raftstore/store/metrics.rs [hunk 9]
```diff
@@ -143,9 +143,23 @@ lazy_static! {
             &["cf"]
         ).unwrap();
 
-    pub static ref STORE_KEYS_WRITTEN_COUNTER: Counter =
+    pub static ref STORE_WRITTEN_KEYS_COUNTER: Counter =
```
Comment (BusyJay) [inline src/raftstore/store/metrics.rs]: I think just `REGION_WRITTEN_KEY_HISTOGRAM` is enough.

File: src/raftstore/store/store.rs (hunk 8 above)
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: This is ok, we don't need exact result.

File: src/raftstore/store/metrics.rs (hunk 9 above)
Comment (siddontang) [inline src/raftstore/store/metrics.rs]: Em, here we may just want to know the written keys rate, if too high, the store may have a high load, so Counter is ok.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/raftstore/store/store.rs [hunk 10]
```diff
@@ -1144,6 +1151,33 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+                peer.written_bytes = 0;
+                peer.written_keys = 0;
+                continue;
+            }
+
+            self.region_written_bytes.observe(peer.written_bytes as f64);
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: Why observe it here rather every time handle ready?

Comment (siddontang) [inline src/raftstore/store/store.rs]: +1， then we can remove the timer here.

Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: `on_raft_ready` know nothing about time,  what we want is to get region flow in a period of time.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**